optuna/integration/pytorch_distributed.py

```diff
@@ -0,0 +1,250 @@
+
+            assert self._delegate is not None
+            return self._delegate.suggest_categorical(name, choices)
+
+        return _call_and_communicate_obj(func, self._device)
+
+    def report(self, value: float, step: int) -> None:
+
+        if dist.get_rank() == 0:
+            assert self._delegate is not None
+            self._delegate.report(value, step)
+        dist.barrier()
+
+    def should_prune(self) -> bool:
+        def func() -> bool:
+
+            assert self._delegate is not None
+            # Some pruners return numpy.bool_, which is incompatible with bool.
+            return bool(self._delegate.should_prune())
+
+        # torch.bool seems to be the correct type, but the communication fails
+        # due to the RuntimeError.
+        return _call_and_communicate(func, torch.uint8, self._device)
+
+    def set_user_attr(self, key: str, value: Any) -> None:
+
+        if dist.get_rank() == 0:
+            assert self._delegate is not None
+            self._delegate.set_user_attr(key, value)
+        dist.barrier()
+
+    def set_system_attr(self, key: str, value: Any) -> None:
+
+        if dist.get_rank() == 0:
+            assert self._delegate is not None
+            self._delegate.set_system_attr(key, value)
+        dist.barrier()
+
+    @property
+    def number(self) -> int:
+        def func() -> int:
+
+            assert self._delegate is not None
+            return self._delegate.number
+
+        return _call_and_communicate(func, torch.int, self._device)
+
+    @property
+    def params(self) -> Dict[str, Any]:
+        def func() -> Dict[str, Any]:
+
+            assert self._delegate is not None
+            return self._delegate.params
+
+        return _call_and_communicate_obj(func, self._device)
+
+    @property
+    def distributions(self) -> Dict[str, BaseDistribution]:
+        def func() -> Dict[str, BaseDistribution]:
+
+            assert self._delegate is not None
+            return self._delegate.distributions
+
+        return _call_and_communicate_obj(func, self._device)
+
+    @property
+    def user_attrs(self) -> Dict[str, Any]:
+        def func() -> Dict[str, Any]:
+
+            assert self._delegate is not None
+            return self._delegate.user_attrs
+
+        return _call_and_communicate_obj(func, self._device)
+
+    @property
+    def system_attrs(self) -> Dict[str, Any]:
+        def func() -> Dict[str, Any]:
+
+            assert self._delegate is not None
+            return self._delegate.system_attrs
+
+        return _call_and_communicate_obj(func, self._device)
+
+    @property
+    def datetime_start(self) -> Optional[datetime]:
+        def func() -> Optional[datetime]:
+
+            assert self._delegate is not None
+            return self._delegate.datetime_start
+
+        return _call_and_communicate_obj(func, self._device)
+
+
+def _call_and_communicate(
+    func: Callable, dtype: "torch.dtype", device: Optional["torch.device"] = None
```

Comment:
Since this function receives `device` arg, which is always `self._device` on the caller side, how about changing it a method (and removing the `device` arg)?
 (if there is no reason you prefer function rather than method)

Reply:
You're absolutely right. I don't have a specific reason to keep it as a function. I'll make `_call_and_communicate` and `_call_and_cummunicate_obj` methods of `TorchDistributedTrial`.

Reply:
I fixed it in the commit https://github.com/optuna/optuna/pull/2303/commits/a03b430cacbd948ee9444c5e6f3adf4a72412cf3.